distributed/scheduler.py

```diff
@@ -7495,16 +7495,28 @@ def decide_worker(
     of bytes sent between workers.  This is determined by calling the
     *objective* function.
     """
+    N_RANDOM_WORKERS: Py_ssize_t = 20
+
     ws: WorkerState = None
     wws: WorkerState
     dts: TaskState
     deps: set = ts._dependencies
+    random_workers_set: set = (
+        valid_workers if valid_workers is not None else all_workers
+    )
     candidates: set
     assert all([dts._who_has for dts in deps])
     if ts._actor:
         candidates = set(all_workers)
     else:
-        candidates = {wws for dts in deps for wws in dts._who_has}
+        if len(random_workers_set) <= N_RANDOM_WORKERS:
+            candidates = random_workers_set
+        else:
+            candidates = {wws for dts in deps for wws in dts._who_has}
+            # TODO the ordering of this set is likely not actually random.
+            # Ideally it would be ordered by occupancy (but that's more work than we want to do).
+            # Can we at least ensure idle workers come first? And that it's reasonably random after that?
```

Comment:
`Scheduler.idle` and `Scheduler.workers` are both SortedDicts, and so should support random access.  You may need to stop using the `workers_dv` attribute, which was brought in for cythonization.  Let's just drop that optimization here, keep things simple, and keep using the sorted dicts.  Then you should be able to do random selections.

Reply:
What are they sorted by?

Reply:
Also, I would first try idle, and then if nothing is there try workers. 

It may be that keeping this second decide_workers out of the Scheduler class isn't worth it.  We shouldn't let the Cythonization effort stop us from making wise scheduling decisions.

Reply:
> What are they sorted by?

Just their name.  Really though it doesn't matter what they're sorted by.  What matters is that these collections are indexable, and so it's possible to do a proper random sample from them in short time.  

You may also want to see how `Scheduler.decide_worker` handles idle workers.

```python
            worker_pool = self._idle or self._workers
            worker_pool_dv = cast(dict, worker_pool)
            wp_vals = worker_pool.values()
            n_workers: Py_ssize_t = len(worker_pool_dv)
            if n_workers < 20:  # smart but linear in small case
                ws = min(wp_vals, key=operator.attrgetter("occupancy"))
                if ws._occupancy == 0:
                    # special case to use round-robin; linear search
                    # for next worker with zero occupancy (or just
                    # land back where we started).
                    wp_i: WorkerState
                    start: Py_ssize_t = self._n_tasks % n_workers
                    i: Py_ssize_t
                    for i in range(n_workers):
                        wp_i = wp_vals[(i + start) % n_workers]
                        if wp_i._occupancy == 0:
                            ws = wp_i
                            break
            else:  # dumb but fast in large case
                ws = wp_vals[self._n_tasks % n_workers]
```

And again, I'm more than happy to walk back the cythonization dict_values / `_dv` stuff if it gets in the way.

Reply:
I was just asking in case what they were sorted by was random enough (or correlated to occupancy in some way) that just iterating through them would be a reasonable way to take the sample. Name is too stable, so I'll use that indexing logic.

Reply:
Yeah, we don't update the sorted collection on every task update.  Live sorting on occupancy would probably be non-viable.  Idle/Saturated is the best we have, and even that ends up taking up non-trivial CPU resources.

Reply:
But generally I think that the decide_worker function probably needs to have a worker pool of `self._idle or self._workers`.  Short term we can probably pass that in as yet-another-parameter.